### PR TITLE
feat(ai-proxy): add native Anthropic Messages API protocol support

### PR DESCRIPTION
**Pull request by nic-6443**
### What this PR does

1. **Native Anthropic Messages protocol adapter** (`ai-protocols/anthropic-messages.lua`)
   - Detects Anthropic requests via URI suffix (`/v1/messages`) + valid JSON body
   - Parses Anthropic SSE events (`message_start`, `content_block_delta`, `message_delta`, `message_stop`)
   - Extracts usage data from Anthropic response format
   - Builds deny responses in Anthropic format
   - Handles message manipulation (content prepend/append)

2. **Bidirectional Anthropic↔OpenAI converter** (`ai-protocols/converters/anthropic-messages-to-openai-chat.lua`)
   - Converts Anthropic requests to OpenAI format (request body, system prompts, tools)
   - Converts OpenAI responses back to Anthropic format (non-streaming and SSE)
   - Handles edge cases: `finish_reason` as JSON null, `usage:null` chunks, empty SSE frames, role+content in first chunk

3. **Protocol registration**
   - Registered `anthropic-messages` in `ai-protocols/init.lua` (detection order: before `openai-chat`)
   - Registered converter in `ai-protocols/converters/init.lua`
   - Added `anthropic-messages` capability to `ai-providers/anthropic.lua`

4. **Refactored `prepare_request` → `prepare_outgoing_request`**
   - Removed `prepare_request` from all protocol modules (was called on client protocol, but `stream_options` injection is a target protocol concern)
   - Inlined model extraction at call site in `ai-proxy/base.lua`
   - Added `prepare_outgoing_request(body)` to `openai-chat.lua` for `stream_options.include_usage=true` injection
   - Called in `ai-providers/base.lua` `build_request()` AFTER conversion, using target protocol module
   - This ensures `stream_options` is correctly injected for both passthrough (OpenAI→OpenAI) and convert (Anthropic→OpenAI) scenarios

5. **Test framework enhancement** (`t/APISIX.pm`)
   - Added `TEST_ENABLE_CONTROL_API_V1` env var support to conditionally disable the `/v1/` control API location block
   - This allows tests using `/v1/messages` URI to work without the control API intercepting requests

6. **Comprehensive test suite** (`t/plugin/ai-proxy-protocol-conversion.t`)
   - 26 test cases covering non-streaming, streaming, error handling, tools, edge cases

### Files changed

**New files:**
- `apisix/plugins/ai-protocols/anthropic-messages.lua`
- `apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua`
- `t/plugin/ai-proxy-protocol-conversion.t`

**Modified files:**
- `apisix/plugins/ai-protocols/init.lua` — register anthropic-messages
- `apisix/plugins/ai-protocols/converters/init.lua` — register converter
- `apisix/plugins/ai-providers/anthropic.lua` — add capability
- `apisix/plugins/ai-protocols/openai-chat.lua` — replace `prepare_request` with `prepare_outgoing_request`
- `apisix/plugins/ai-protocols/openai-embeddings.lua` — remove `prepare_request`
- `apisix/plugins/ai-providers/base.lua` — add `prepare_outgoing_request` call after conversion
- `apisix/plugins/ai-proxy/base.lua` — inline model extraction, store `ctx.ai_target_protocol`
- `t/APISIX.pm` — add `TEST_ENABLE_CONTROL_API_V1` env var support

### CODE REVIEW

File: apisix/plugins/ai-protocols/anthropic-messages.lua [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+            .. "\n"
+            .. sse.encode({ type = "content_block_delta",
+                            data = core.json.encode(content_block_delta) })
+            .. "\n"
+            .. sse.encode({ type = "content_block_stop",
+                            data = core.json.encode(content_block_stop) })
+            .. "\n"
+            .. sse.encode({ type = "message_delta",
+                            data = core.json.encode(message_delta) })
+            .. "\n"
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-protocols/anthropic-messages.lua]: Good catch. Fixed in 09ad2ff — removed the redundant newline separators between sse.encode() calls since sse.encode() already terminates each event with \n\n.

File: apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+                }
+            })
+        end
+        openai_body.tools = openai_tools
+    end
+
+    -- 4. Map Parameters
+    if openai_body.max_tokens then
+        openai_body.max_completion_tokens = openai_body.max_tokens
+    end
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua:210]: This mapping is intentional. The OpenAI API has deprecated `max_tokens` in favor of `max_completion_tokens` for newer models (o1, o3, gpt-4o, etc.). The Anthropic API uses `max_tokens` to mean "max output tokens", which aligns semantically with `max_completion_tokens` rather than the old `max_tokens` (which in early OpenAI API versions had ambiguous scope). Most OpenAI-compatible providers accept both fields, and this ensures correct behavior with newer model endpoints.

File: apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+
+    if openai_body.stop_sequences then
+        openai_body.stop = openai_body.stop_sequences
+        openai_body.stop_sequences = nil
+    end
+
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua:216]: Good catch! Fixed in 09ad2ff — now injecting `stream_options = { include_usage = true }` in the converter when `openai_body.stream` is true. This ensures OpenAI-compatible providers include usage data in streaming responses.

File: apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+        if type(msg.content) == "string" then
+            new_msg.content = msg.content
+        elseif type(msg.content) == "table" then
+            local tool_calls = {}
+            local tool_results = {}
+
+            for _, block in ipairs(msg.content) do
+                if type(block) ~= "table" then
+                    core.log.warn("unexpected non-table content block in Anthropic ",
+                                  "response, skipping: ", tostring(block))
```
[the remark this answers is not in the document]
Comment (nic-6443) [reply apisix/plugins/ai-protocols/converters/anthropic-messages-to-openai-chat.lua]: Good catch. Fixed in 09ad2ff — updated the log message from "response" to "request" since this is in `convert_request()`.

File: t/APISIX.pm [hunk 5]
```diff
@@ -777,6 +777,22 @@ _EOC_
         $ipv6_listen_conf = "listen \[::1\]:1984;"
     }
 
+    my $enable_test_control_api_v1 =
+        !defined($ENV{TEST_ENABLE_CONTROL_API_V1}) ||
+        $ENV{TEST_ENABLE_CONTROL_API_V1} ne "0";
+
+    my $control_api_v1_location = "";
```
Comment (shreemaan-abhishek) [inline t/APISIX.pm:784]: why do we need these changes?

Comment (nic-6443) [reply t/APISIX.pm:784]: The test framework has a hardcoded `location /v1/` block for the control API (`apisix.http_control()`), which intercepts **all** `/v1/*` requests. The Anthropic Messages API uses `/v1/messages` as its endpoint, so without this change, all test requests to `/v1/messages` are intercepted by the control API location block and return 404 instead of reaching the mock upstream.

This `TEST_ENABLE_CONTROL_API_V1` env var allows tests that use `/v1/*` paths (like the Anthropic protocol tests added in this PR) to conditionally disable the control API location. When a test file sets `$ENV{TEST_ENABLE_CONTROL_API_V1} = "0"` in its `BEGIN` block, the `/v1/` location is excluded from the nginx config, allowing `/v1/messages` requests to pass through to the upstream normally.

The test file `t/plugin/ai-proxy-protocol-conversion.t` uses this mechanism — see its `BEGIN` block.

**Review by AlinsRan — APPROVED** [review]

**Review by membphis — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]
